packages/apps/human-app/frontend/src/shared/contexts/generic-auth-context.tsx

```diff
@@ -0,0 +1,152 @@
+import { type AuthTokensSuccessResponse } from '@/shared/schemas';
+import { type UserData } from '@/modules/auth/context/auth-context';
+import { type Web3UserData } from '@/modules/auth-web3/context/web3-auth-context';
+import { type AuthType } from '../types/browser-auth-provider';
+
+export type AuthStatus = 'loading' | 'error' | 'success' | 'idle';
+
+interface AuthContextType<T> {
+  status: AuthStatus;
+  signOut: ({
+    throwExpirationModal,
+    status,
+  }: {
+    throwExpirationModal?: boolean;
+    status?: AuthStatus;
+  }) => void;
+  signIn: (signInSuccess: AuthTokensSuccessResponse) => void;
+  updateUserData: (update: Partial<T>) => void;
+}
+
+export interface AuthenticatedUserContextType<T> extends AuthContextType<T> {
+  user: T;
+}
+
+export interface UnauthenticatedUserContextType<T> extends AuthContextType<T> {
+  user: null;
+}
+
+export function createAuthProvider<T extends UserData | Web3UserData>(config: {
+  authType: AuthType;
+  schema: ZodType<T>;
+}) {
+  const { authType, schema } = config;
+  const AuthContext = createContext<
+    AuthenticatedUserContextType<T> | UnauthenticatedUserContextType<T> | null
+  >(null);
+
+  function AuthProvider({ children }: Readonly<{ children: React.ReactNode }>) {
+    const queryClient = useQueryClient();
+    const { openModal } = useModalStore();
+    const [authState, setAuthState] = useState<{
+      user: T | null;
+      status: AuthStatus;
+    }>({
+      user: null,
+      status: 'loading',
+    });
+
+    const displayExpirationModal = () => {
+      queryClient.setDefaultOptions({ queries: { enabled: false } });
+      openModal({
+        modalType: ModalType.EXPIRATION_MODAL,
+        displayCloseButton: false,
+        maxWidth: 'sm',
+      });
+    };
+
+    const handleSignIn = () => {
+      try {
+        const accessToken = browserAuthProvider.getAccessToken();
+        const currentAuthType = browserAuthProvider.getAuthType();
+        const savedUserData = browserAuthProvider.getUserData();
+
+        if (!accessToken || currentAuthType !== authType) {
+          setAuthState({ user: null, status: 'idle' });
+          return;
+        }
+        const userData = jwtDecode(accessToken);
+        const userDataWithSavedData = savedUserData.data
+          ? { ...userData, ...savedUserData.data }
+          : userData;
+
+        const validUserData = schema.parse(userDataWithSavedData);
+
+        setAuthState({ user: validUserData, status: 'success' });
+
+        browserAuthProvider.signOutSubscription = displayExpirationModal;
+      } catch (e) {
+        // eslint-disable-next-line no-console
+        console.error('Invalid JWT payload:', e);
+        signOut({
+          status: 'error',
+        });
+      }
+    };
+
+    const signIn = (signInSuccess: AuthTokensSuccessResponse) => {
+      browserAuthProvider.signIn(signInSuccess, authType);
+      handleSignIn();
+    };
+
+    const signOut = ({
```

Comment:
This function has new signature now and because of that build fails since it's called with invalid args in different places.
Don't forget to provide proper defaults when you fix it